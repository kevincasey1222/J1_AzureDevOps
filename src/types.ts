import { IntegrationInstanceConfig } from '@jupiterone/integration-sdk-core';

/**
 * Properties provided by the `IntegrationInstance.config`. This reflects the
 * same properties defined by `instanceConfigFields`.
 */
export interface IntegrationConfig extends IntegrationInstanceConfig {
  // clientId: string; // not used
  // clientSecret: string; // not used
  /**
   * The provider API client ID used to authenticate requests.
   */
  orgUrl: string;

  /**
   * The personal access token used to authenticate requests.
   */
  accessToken: string;
}
