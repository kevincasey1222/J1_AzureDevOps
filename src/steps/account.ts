import {
  createIntegrationEntity,
  IntegrationStep,
  IntegrationStepExecutionContext,
} from '@jupiterone/integration-sdk-core';

import { ADOIntegrationConfig } from '../types';

export const AZURE_DEVOPS_ACCOUNT = 'azure_devops_account';

export async function fetchAccountDetails({
  instance,
  jobState,
}: IntegrationStepExecutionContext<ADOIntegrationConfig>) {
  const name = `Azure Devops - ${instance.name}`;
  const accountEntity = await jobState.addEntity(
    createIntegrationEntity({
      entityData: {
        source: {
          id: 'azure-devops',
          name: 'Azure Devops Account',
        },
        assign: {
          _key: `azure-devops-account:${instance.id}`,
          _type: AZURE_DEVOPS_ACCOUNT,
          _class: 'Account',
          name,
          displayName: name,
        },
      },
    }),
  );
  await jobState.setData(AZURE_DEVOPS_ACCOUNT, accountEntity);
}

export const accountSteps: IntegrationStep<ADOIntegrationConfig>[] = [
  {
    id: 'fetch-account',
    name: 'Fetch Account Details',
    entities: [
      {
        resourceName: 'Azure Devops Account',
        _type: AZURE_DEVOPS_ACCOUNT,
        _class: 'Account',
      },
    ],
    relationships: [],
    dependsOn: [],
    executionHandler: fetchAccountDetails,
  },
];
