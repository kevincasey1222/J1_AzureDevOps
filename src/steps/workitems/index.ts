import {
  createDirectRelationship,
  createIntegrationEntity,
  Entity,
  IntegrationStep,
  IntegrationStepExecutionContext,
  RelationshipClass,
  IntegrationMissingKeyError,
} from '@jupiterone/integration-sdk-core';

import { createAPIClient } from '../../client';
import { ADOIntegrationConfig } from '../../types';

export async function fetchWorkitems({
  instance,
  jobState,
}: IntegrationStepExecutionContext<ADOIntegrationConfig>) {
  const apiClient = createAPIClient(instance.config);

  await apiClient.iterateWorkitems(async (item) => {
    const workItemEntity = await jobState.addEntity(
      createIntegrationEntity({
        entityData: {
          source: item,
          assign: {
            _type: 'azure_devops_work_item',
            _class: 'Record',
            _key: (item.projectId || '') + item.id?.toString(),
            name: (item.fields || {})['System.Title'],
            displayName: (item.fields || {})['System.Title'],
            type: (item.fields || {})['System.WorkItemType'],
            webLink: item.url,
            workItemType: (item.fields || {})['System.WorkItemType'],
            description: (item.fields || {})['System.Description'],
            projectId: item.projectId,
            projectName: (item.fields || {})['System.TeamProject'],
            teamProject: (item.fields || {})['System.TeamProject'],
            revision: item.rev,
            areaPath: (item.fields || {})['System.AreaPath'],
            interationPath: (item.fields || {})['System.IterationPath'],
            state: (item.fields || {})['System.State'],
            reason: (item.fields || {})['System.Reason'],
            createdDate: (item.fields || {})['System.CreatedDate'],
            createdBy: (item.fields || {})['System.CreatedBy'],
            changedDate: (item.fields || {})['System.ChangedDate'],
            changedBy: (item.fields || {})['System.ChangedBy'],
            commentCount: (item.fields || {})['System.CommentCount'],
            stateChangeDate: (item.fields || {})[
              'Microsoft.VSTS.Common.StateChangeDate'
            ],
            priority: (item.fields || {})['Microsoft.VSTS.Common.Priority'],
            history: (item.fields || {})['System.History'],
          },
        },
      }),
    );

    if (item.projectId != undefined) {
      const projectEntity = await jobState.findEntity(item.projectId);
      if (!projectEntity) {
        throw new IntegrationMissingKeyError(
          `Expected project with key to exist (key=${item.projectId})`,
        );
      }
      await jobState.addRelationship(
        createDirectRelationship({
          _class: RelationshipClass.HAS,
          from: projectEntity,
          to: workItemEntity,
        }),
      );
    }
  });
}

export const workitemSteps: IntegrationStep<ADOIntegrationConfig>[] = [
  {
    id: 'fetch-workitems',
    name: 'Fetch Workitems',
    entities: [
      {
        resourceName: 'ADO WorkItem',
        _type: 'azure_devops_work_item',
        _class: 'Record',
      },
    ],
    relationships: [
      {
        _type: 'azure_devops_project_has_work_item',
        _class: RelationshipClass.HAS,
        sourceType: 'azure_devops_project',
        targetType: 'azure_devops_work_item',
      },
    ],
    dependsOn: ['fetch-projects'],
    executionHandler: fetchWorkitems,
  },
];
