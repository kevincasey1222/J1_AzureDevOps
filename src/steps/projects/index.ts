import {
  createDirectRelationship,
  createIntegrationEntity,
  Entity,
  IntegrationStep,
  IntegrationStepExecutionContext,
  RelationshipClass,
  //IntegrationMissingKeyError,
} from '@jupiterone/integration-sdk-core';

import { createAPIClient } from '../../client';
import { ADOIntegrationConfig } from '../../types';
import { AZURE_DEVOPS_ACCOUNT } from '../account';

export async function fetchProjects({
  instance,
  jobState,
}: IntegrationStepExecutionContext<ADOIntegrationConfig>) {
  const apiClient = createAPIClient(instance.config);

  const accountEntity = (await jobState.getData(
    AZURE_DEVOPS_ACCOUNT,
  )) as Entity;

  await apiClient.iterateProjects(async (project) => {
    const projectEntity = await jobState.addEntity(
      createIntegrationEntity({
        entityData: {
          source: project,
          assign: {
            _type: 'azure_devops_project',
            _class: 'Project',
            _key: project.id,
            name: project.name,
            displayName: project.name,
            abbreviation: project.abbreviation,
            description: project.description,
            webLink: project.url,
            state: project.state,
            revision: project.revision,
            visibility: project.visibility,
            createdOn: undefined,
            updatedOn: undefined,
          },
        },
      }),
    );

    await jobState.addRelationship(
      createDirectRelationship({
        _class: RelationshipClass.HAS,
        from: accountEntity,
        to: projectEntity,
      }),
    );
  });
}

export const projectSteps: IntegrationStep<ADOIntegrationConfig>[] = [
  {
    id: 'fetch-projects',
    name: 'Fetch Projects',
    entities: [
      {
        resourceName: 'ADO Project',
        _type: 'azure_devops_project',
        _class: 'Project',
      },
    ],
    relationships: [
      {
        _type: 'azure_devops_account_has_project',
        _class: RelationshipClass.HAS,
        sourceType: 'azure_devops_account',
        targetType: 'azure_devops_project',
      },
    ],
    dependsOn: ['fetch-account'],
    executionHandler: fetchProjects,
  },
];
