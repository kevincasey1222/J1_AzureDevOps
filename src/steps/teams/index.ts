import {
  createDirectRelationship,
  createIntegrationEntity,
  Entity,
  IntegrationStep,
  IntegrationStepExecutionContext,
  RelationshipClass,
  IntegrationMissingKeyError,
} from '@jupiterone/integration-sdk-core';
//import { ResultGroupType } from 'azure-devops-node-api/interfaces/TestInterfaces';

import { createAPIClient } from '../../client';
import { ADOIntegrationConfig } from '../../types';
import { AZURE_DEVOPS_ACCOUNT } from '../account';

export async function fetchTeams({
  instance,
  jobState,
}: IntegrationStepExecutionContext<ADOIntegrationConfig>) {
  const apiClient = createAPIClient(instance.config);

  const accountEntity = (await jobState.getData(
    AZURE_DEVOPS_ACCOUNT,
  )) as Entity;

  await apiClient.iterateGroups(async (team) => {
    const groupEntity = await jobState.addEntity(
      createIntegrationEntity({
        entityData: {
          source: team,
          assign: {
            _type: 'azure_devops_team',
            _class: 'UserGroup',
            _key: team.id,
            name: team.name,
            displayName: team.name,
            webLink: team.url,
            description: team.description,
            identityUrl: team.identityUrl,
            projectName: team.projectName,
            projectId: team.projectId,
          },
        },
      }),
    );

    await jobState.addRelationship(
      createDirectRelationship({
        _class: RelationshipClass.HAS,
        from: accountEntity,
        to: groupEntity,
      }),
    );

    for (const user of team.users || []) {
      const userEntity = await jobState.findEntity(user.id);

      if (!userEntity) {
        throw new IntegrationMissingKeyError(
          `Expected user with key to exist (key=${user.id})`,
        );
      }

      await jobState.addRelationship(
        createDirectRelationship({
          _class: RelationshipClass.HAS,
          from: groupEntity,
          to: userEntity,
        }),
      );
    }

    if (team.projectId != undefined) {
      const projectEntity = await jobState.findEntity(team.projectId);
      if (!projectEntity) {
        throw new IntegrationMissingKeyError(
          `Expected project with key to exist (key=${team.projectId})`,
        );
      }
      await jobState.addRelationship(
        createDirectRelationship({
          _class: RelationshipClass.HAS,
          from: projectEntity,
          to: groupEntity,
        }),
      );
    }
    
  });
}

export const teamSteps: IntegrationStep<ADOIntegrationConfig>[] = [
  {
    id: 'fetch-teams',
    name: 'Fetch Teams',
    entities: [
      {
        resourceName: 'ADO Team',
        _type: 'azure_devops_team',
        _class: 'UserGroup',
      },
    ],
    relationships: [
      {
        _type: 'azure_devops_project_has_team',
        _class: RelationshipClass.HAS,
        sourceType: 'azure_devops_project',
        targetType: 'azure_devops_team',
      },
      {
        _type: 'azure_devops_account_has_team',
        _class: RelationshipClass.HAS,
        sourceType: 'azure_devops_account',
        targetType: 'azure_devops_team',
      },
      {
        _type: 'azure_devops_team_has_user',
        _class: RelationshipClass.HAS,
        sourceType: 'azure_devops_team',
        targetType: 'azure_devops_user',
      },
    ],
    dependsOn: ['fetch-users'],
    executionHandler: fetchTeams,
  },
];
