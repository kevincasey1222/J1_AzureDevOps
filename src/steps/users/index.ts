import {
  createDirectRelationship,
  createIntegrationEntity,
  Entity,
  IntegrationStep,
  IntegrationStepExecutionContext,
  RelationshipClass,
  //IntegrationMissingKeyError,
} from '@jupiterone/integration-sdk-core';
//import { ResultGroupType } from 'azure-devops-node-api/interfaces/TestInterfaces';

import { createAPIClient } from '../../client';
import { ADOIntegrationConfig } from '../../types';
import { AZURE_DEVOPS_ACCOUNT } from '../account';

export async function fetchUsers({
  instance,
  jobState,
}: IntegrationStepExecutionContext<ADOIntegrationConfig>) {
  const apiClient = createAPIClient(instance.config);

  const accountEntity = (await jobState.getData(
    AZURE_DEVOPS_ACCOUNT,
  )) as Entity;

  await apiClient.iterateUsers(async (user) => {
    const userEntity = await jobState.addEntity(
      createIntegrationEntity({
        entityData: {
          source: user,
          assign: {
            _type: 'azure_devops_user',
            _class: 'User',
            _key: user.identity?.id,
            name: user.identity?.displayName,
            displayName: user.identity?.displayName,
            email: user.identity?.uniqueName,
            username: user.identity?.uniqueName,
            webLink: user.identity?.url,
            imageLink: user.identity?.imageUrl,
            description: user.identity?.descriptor,
            profileLink: user.identity?.profileUrl,
          },
        },
      }),
    );

    await jobState.addRelationship(
      createDirectRelationship({
        _class: RelationshipClass.HAS,
        from: accountEntity,
        to: userEntity,
      }),
    );
  });
}

export const userSteps: IntegrationStep<ADOIntegrationConfig>[] = [
  {
    id: 'fetch-users',
    name: 'Fetch Users',
    entities: [
      {
        resourceName: 'ADO User',
        _type: 'azure_devops_user',
        _class: 'User',
      },
    ],
    relationships: [
      {
        _type: 'azure_devops_account_has_user',
        _class: RelationshipClass.HAS,
        sourceType: 'azure_devops_account',
        targetType: 'azure_devops_user',
      },
    ],
    dependsOn: ['fetch-account'],
    executionHandler: fetchUsers,
  },
];
