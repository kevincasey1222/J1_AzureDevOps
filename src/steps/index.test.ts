import { createMockStepExecutionContext } from '@jupiterone/integration-sdk-testing';

import { ADOIntegrationConfig } from '../types';
import { fetchProjects } from './projects';
import { fetchWorkitems } from './workitems';
import { fetchTeams } from './teams';
import { fetchUsers } from './users';
import { fetchAccountDetails } from './account';

const DEFAULT_ORG_URL = 'https://dev.azure.com/tkcasey1';
const DEFAULT_ACCESS_TOKEN =
  'vfpvjk6lggyfu5bcglvenyhfllyitn2toalf33srqdhmgnqdykra';

const integrationConfig: ADOIntegrationConfig = {
  orgUrl: process.env.ORG_URL || DEFAULT_ORG_URL,
  accessToken: process.env.ACCESS_TOKEN || DEFAULT_ACCESS_TOKEN,
};

jest.setTimeout(25000);

test('should collect data', async () => {
  const context = createMockStepExecutionContext<ADOIntegrationConfig>({
    instanceConfig: integrationConfig,
  });

  // Simulates dependency graph execution.
  // See https://github.com/JupiterOne/sdk/issues/262.
  await fetchAccountDetails(context);
  await fetchProjects(context);
  await fetchWorkitems(context);
  await fetchUsers(context);
  await fetchTeams(context);

  console.log(context.jobState.collectedEntities.length);
  console.log(context.jobState.collectedRelationships.length);
  console.log(context.jobState.collectedEntities);
  console.log(context.jobState.collectedRelationships);
  console.log(context.jobState.encounteredTypes);

  // Review snapshot, failure is a regression
  expect({
    numCollectedEntities: context.jobState.collectedEntities.length,
    numCollectedRelationships: context.jobState.collectedRelationships.length,
    collectedEntities: context.jobState.collectedEntities,
    collectedRelationships: context.jobState.collectedRelationships,
    encounteredTypes: context.jobState.encounteredTypes,
  }).toMatchSnapshot();

  const accounts = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('Account'),
  );
  expect(accounts.length).toBeGreaterThan(0);
  expect(accounts).toMatchGraphObjectSchema({
    _class: ['Account'],
    schema: {
      additionalProperties: false,
      properties: {
        _type: { const: 'azure_devops_account' },
        _key: { type: 'string' },
        name: { type: 'string' },
        displayName: { type: 'string' },
      },
      required: ['name'],
    },
  });

  /*
  const users = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('User'),
  );
  expect(users.length).toBeGreaterThan(0);
  expect(users).toMatchGraphObjectSchema({
    _class: ['User'],
    schema: {
      additionalProperties: false,
      properties: {
        _type: { const: 'acme_user' },
        firstName: { type: 'string' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: ['firstName'],
    },
  });

  const userGroups = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('UserGroup'),
  );
  expect(userGroups.length).toBeGreaterThan(0);
  expect(userGroups).toMatchGraphObjectSchema({
    _class: ['UserGroup'],
    schema: {
      additionalProperties: false,
      properties: {
        _type: { const: 'acme_group' },
        logoLink: {
          type: 'string',
          // Validate that the `logoLink` property has a URL format
          format: 'url',
        },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: ['logoLink'],
    },
  }); */
});
