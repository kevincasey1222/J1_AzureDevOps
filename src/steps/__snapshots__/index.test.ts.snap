// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "azure-devops-account:local-integration-instance",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": "azure-devops",
            "name": "Azure Devops Account",
          },
        },
      ],
      "_type": "azure_devops_account",
      "createdOn": undefined,
      "displayName": "Azure Devops - Local Integration",
      "id": "azure-devops",
      "name": "Azure Devops - Local Integration",
    },
    Object { 
      id: '46464f21-a350-4ade-8a9d-cdac1b959b21',
      name: 'Project Beta',
      description: 'This project isn\'t really quite as cool, but it\'s alright.',
      _type: 'azure_devops_project', 
      _class: [ 'Project' ],
      _key: '46464f21-a350-4ade-8a9d-cdac1b959b21',
      displayName: 'Project Beta',  
      abbreviation: undefined, 
      webLink: 'https://dev.azure.com/tkcasey1/_apis/projects/46464f21-a350-4ade-8a9d-cdac1b959b21',
      state: 'wellFormed',
      revision: 20,
      visibility: 0,  
      createdOn: undefined, 
      updatedOn: undefined,  
      _rawData: [ [Object] ] 
    },
    Object { 
      id: 'd5ec9e77-afa6-45c9-a2a5-2e015714f4fa', 
      name: 'Initial Project', 
      description: 'This is a great project. No, really, it is.', 
      _type: 'azure_devops_project',
      _class: [ 'Project' ], 
      _key: 'd5ec9e77-afa6-45c9-a2a5-2e015714f4fa',
      displayName: 'Initial Project', 
      abbreviation: undefined, 
      webLink: 'https://dev.azure.com/tkcasey1/_apis/projects/d5ec9e77-afa6-45c9-a2a5-2e015714f4fa',
      state: 'wellFormed',
      revision: 12, 
      visibility: 0, 
      createdOn: undefined, 
      updatedOn: undefined,
      _rawData: [ [Object] ] 
    },
    Object { 
      _type: 'azure_devops_work_item', 
      _class: [ 'Record' ], 
      _key: 'd5ec9e77-afa6-45c9-a2a5-2e015714f4fa1',
      name: 'Test Epic',
      displayName: 'Test Epic', 
      type: 'Epic', 
      webLink: undefined, 
      workItemType: 'Epic', 
      description: undefined,
      projectId: 'd5ec9e77-afa6-45c9-a2a5-2e015714f4fa', 
      projectName: 'Initial Project',
      teamProject: 'Initial Project', 
      revision: 1, 
      areaPath: 'Initial Project',  
      interationPath: 'Initial Project',  
      state: 'To Do', 
      reason: 'Added to backlog',
      createdDate: '2020-11-17T21:18:40.36Z', 
      createdBy: 'Kevin Casey <tkcasey1@sundevils.asu.edu>',  
      changedDate: '2020-11-17T21:18:40.36Z',
      changedBy: 'Kevin Casey <tkcasey1@sundevils.asu.edu>',
      commentCount: 1,
      stateChangeDate: '2020-11-17T21:18:40.36Z',
      priority: 2,
      history: undefined,  
      _rawData: [ [Object] ], 
      createdOn: undefined 
    },
    Object { 
      _type: 'azure_devops_work_item',
      _class: [ 'Record' ], 
      _key: 'd5ec9e77-afa6-45c9-a2a5-2e015714f4fa2',
      name: 'Top Level Task',
      displayName: 'Top Level Task',
      type: 'Task',
      webLink: undefined,
      workItemType: 'Task', 
      description: undefined,
      projectId: 'd5ec9e77-afa6-45c9-a2a5-2e015714f4fa',
      projectName: 'Initial Project',
      teamProject: 'Initial Project',
      revision: 1,
      areaPath: 'Initial Project',
      interationPath: 'Initial Project',
      state: 'To Do',
      reason: 'Added to backlog',
      createdDate: '2020-11-17T21:19:19.85Z',
      createdBy: 'Kevin Casey <tkcasey1@sundevils.asu.edu>',
      changedDate: '2020-11-17T21:19:19.85Z', 
      changedBy: 'Kevin Casey <tkcasey1@sundevils.asu.edu>',
      commentCount: 0, 
      stateChangeDate: '2020-11-17T21:19:19.85Z', 
      priority: 2, 
      history: undefined, 
      _rawData: [ [Object] ],
      createdOn: undefined 
    },
    Object { 
      _type: 'azure_devops_user',
      _class: [ 'User' ], 
      _key: 'ba1904ec-e70b-6a46-9949-131ecb022f58', 
      name: 'Kevin Casey',
      displayName: 'Kevin Casey',
      email: 'tkcasey1@sundevils.asu.edu', 
      username: 'tkcasey1@sundevils.asu.edu',
      webLink: 'https://spsprodcus4.vssps.visualstudio.com/Ab7805cb5-c11a-492d-a52f-d8e15210042a/_apis/Identities/ba1904ec-e70b-6a46-9949-131ecb022f58',
      imageLink: 'https://dev.azure.com/tkcasey1/_api/_common/identityImage?id=ba1904ec-e70b-6a46-9949-131ecb022f58',
      description: 'aad.YmExOTA0ZWMtZTcwYi03YTQ2LTk5NDktMTMxZWNiMDIyZjU4',  
      profileLink: undefined,
      _rawData: [ [Object] ],
      createdOn: undefined 
      },
    Object { 
      _type: 'azure_devops_user',
      _class: [ 'User' ],
      _key: 'a6374253-7f1b-42fd-9981-30fe0d272182',
      name: 'Thomas Luce',
      displayName: 'Thomas Luce',
      email: 'thomas.luce@gmail.com',
      username: 'thomas.luce@gmail.com',
      webLink: 'https://spsprodcus4.vssps.visualstudio.com/Ab7805cb5-c11a-492d-a52f-d8e15210042a/_apis/Identities/a6374253-7f1b-42fd-9981-30fe0d272182',
      imageLink: 'https://dev.azure.com/tkcasey1/_api/_common/identityImage?id=a6374253-7f1b-42fd-9981-30fe0d272182',
      description: 'aad.YjUzMDBmMTMtODc3OC03NWQwLWEyMWQtNTRjMGJiZjAxNDQz', 
      profileLink: undefined,
      _rawData: [ [Object] ],
      createdOn: undefined 
    },
    Object { 
      id: '7dba04a8-12a9-49ae-b646-77b6fe11bf53',
      name: 'Initial Project Team', 
      description: 'The default project team.',
      _type: 'azure_devops_team',
      _class: [ 'UserGroup' ], 
      _key: '7dba04a8-12a9-49ae-b646-77b6fe11bf53',
      displayName: 'Initial Project Team',
      webLink: 'https://dev.azure.com/tkcasey1/_apis/projects/d5ec9e77-afa6-45c9-a2a5-2e015714f4fa/teams/7dba04a8-12a9-49ae-b646-77b6fe11bf53',
      identityUrl: 'https://spsprodcus4.vssps.visualstudio.com/Ab7805cb5-c11a-492d-a52f-d8e15210042a/_apis/Identities/7dba04a8-12a9-49ae-b646-77b6fe11bf53',
      projectName: 'Initial Project',
      projectId: 'd5ec9e77-afa6-45c9-a2a5-2e015714f4fa',
      _rawData: [ [Object] ], 
      createdOn: undefined 
    },
    Object { 
      id: '68646bd1-db9c-4e38-9d5f-09eb990fc0b9',
      name: 'Project Beta Team',
      description: 'The default project team.', 
      _type: 'azure_devops_team',
      _class: [ 'UserGroup' ], 
      _key: '68646bd1-db9c-4e38-9d5f-09eb990fc0b9', 
      displayName: 'Project Beta Team', 
      webLink: 'https://dev.azure.com/tkcasey1/_apis/projects/46464f21-a350-4ade-8a9d-cdac1b959b21/teams/68646bd1-db9c-4e38-9d5f-09eb990fc0b9',
      identityUrl: 'https://spsprodcus4.vssps.visualstudio.com/Ab7805cb5-c11a-492d-a52f-d8e15210042a/_apis/Identities/68646bd1-db9c-4e38-9d5f-09eb990fc0b9',
      projectName: 'Project Beta',
      projectId: '46464f21-a350-4ade-8a9d-cdac1b959b21',
      _rawData: [ [Object] ], 
      createdOn: undefined 
    },    
  ],
  "collectedRelationships": Array [
    Object { 
      _key: 'azure-devops-account:local-integration-instance|has|46464f21-a350-4ade-8a9d-cdac1b959b21',
      _type: 'azure_devops_account_has_project',
      _class: 'HAS',
      _fromEntityKey: 'azure-devops-account:local-integration-instance',
      _toEntityKey: '46464f21-a350-4ade-8a9d-cdac1b959b21',
      displayName: 'HAS' 
    },
    Object { 
      _key: 'azure-devops-account:local-integration-instance|has|d5ec9e77-afa6-45c9-a2a5-2e015714f4fa',
      _type: 'azure_devops_account_has_project', 
      _class: 'HAS',
      _fromEntityKey: 'azure-devops-account:local-integration-instance', 
      _toEntityKey: 'd5ec9e77-afa6-45c9-a2a5-2e015714f4fa', 
      displayName: 'HAS' 
    },
    Object { 
      _key: 'd5ec9e77-afa6-45c9-a2a5-2e015714f4fa|has|d5ec9e77-afa6-45c9-a2a5-2e015714f4fa1',
      _type: 'azure_devops_project_has_work_item',
      _class: 'HAS',     
      _fromEntityKey: 'd5ec9e77-afa6-45c9-a2a5-2e015714f4fa',
      _toEntityKey: 'd5ec9e77-afa6-45c9-a2a5-2e015714f4fa1', 
      displayName: 'HAS' 
    },
    Object { 
      _key: 'd5ec9e77-afa6-45c9-a2a5-2e015714f4fa|has|d5ec9e77-afa6-45c9-a2a5-2e015714f4fa2',
      _type: 'azure_devops_project_has_work_item',  
      _class: 'HAS', 
      _fromEntityKey: 'd5ec9e77-afa6-45c9-a2a5-2e015714f4fa', 
      _toEntityKey: 'd5ec9e77-afa6-45c9-a2a5-2e015714f4fa2', 
      displayName: 'HAS' 
    }, 
    Object { 
      _key: 'azure-devops-account:local-integration-instance|has|ba1904ec-e70b-6a46-9949-131ecb022f58',
      _type: 'azure_devops_account_has_user',
      _class: 'HAS',
      _fromEntityKey: 'azure-devops-account:local-integration-instance',
      _toEntityKey: 'ba1904ec-e70b-6a46-9949-131ecb022f58', 
      displayName: 'HAS' 
    },
    Object { 
      _key: 'azure-devops-account:local-integration-instance|has|a6374253-7f1b-42fd-9981-30fe0d272182',
      _type: 'azure_devops_account_has_user', 
      _class: 'HAS', 
      _fromEntityKey: 'azure-devops-account:local-integration-instance', 
      _toEntityKey: 'a6374253-7f1b-42fd-9981-30fe0d272182', 
      displayName: 'HAS' 
    },
    Object { 
      _key: 'azure-devops-account:local-integration-instance|has|7dba04a8-12a9-49ae-b646-77b6fe11bf53',
      _type: 'azure_devops_account_has_team',
      _class: 'HAS', 
      _fromEntityKey: 'azure-devops-account:local-integration-instance',
      _toEntityKey: '7dba04a8-12a9-49ae-b646-77b6fe11bf53', 
      displayName: 'HAS' 
    },
    Object { 
      _key: '7dba04a8-12a9-49ae-b646-77b6fe11bf53|has|ba1904ec-e70b-6a46-9949-131ecb022f58',
      _type: 'azure_devops_team_has_user',
      _class: 'HAS',
      _fromEntityKey: '7dba04a8-12a9-49ae-b646-77b6fe11bf53',
      _toEntityKey: 'ba1904ec-e70b-6a46-9949-131ecb022f58',
      displayName: 'HAS' 
    },
    Object { 
      _key: '7dba04a8-12a9-49ae-b646-77b6fe11bf53|has|a6374253-7f1b-42fd-9981-30fe0d272182',
      _type: 'azure_devops_team_has_user',
      _class: 'HAS',
      _fromEntityKey: '7dba04a8-12a9-49ae-b646-77b6fe11bf53',
      _toEntityKey: 'a6374253-7f1b-42fd-9981-30fe0d272182',
      displayName: 'HAS' 
    },
    Object { 
      _key: 'd5ec9e77-afa6-45c9-a2a5-2e015714f4fa|has|7dba04a8-12a9-49ae-b646-77b6fe11bf53',
      _type: 'azure_devops_project_has_team', 
      _class: 'HAS',
      _fromEntityKey: 'd5ec9e77-afa6-45c9-a2a5-2e015714f4fa',
      _toEntityKey: '7dba04a8-12a9-49ae-b646-77b6fe11bf53',   
      displayName: 'HAS' 
    },
    Object { 
      _key: 'azure-devops-account:local-integration-instance|has|68646bd1-db9c-4e38-9d5f-09eb990fc0b9',
      _type: 'azure_devops_account_has_team',
      _class: 'HAS',
      _fromEntityKey: 'azure-devops-account:local-integration-instance',
      _toEntityKey: '68646bd1-db9c-4e38-9d5f-09eb990fc0b9',
      displayName: 'HAS' 
    },
    Object { 
      _key: '68646bd1-db9c-4e38-9d5f-09eb990fc0b9|has|ba1904ec-e70b-6a46-9949-131ecb022f58',
      _type: 'azure_devops_team_has_user',
      _class: 'HAS',
      _fromEntityKey: '68646bd1-db9c-4e38-9d5f-09eb990fc0b9', 
      _toEntityKey: 'ba1904ec-e70b-6a46-9949-131ecb022f58',
      displayName: 'HAS' 
    },
    Object { 
      _key: '46464f21-a350-4ade-8a9d-cdac1b959b21|has|68646bd1-db9c-4e38-9d5f-09eb990fc0b9',
      _type: 'azure_devops_project_has_team', 
      _class: 'HAS',
      _fromEntityKey: '46464f21-a350-4ade-8a9d-cdac1b959b21',
      _toEntityKey: '68646bd1-db9c-4e38-9d5f-09eb990fc0b9',
      displayName: 'HAS' 
    }  
  ],
  "encounteredTypes": Array [
    'azure_devops_account',
    'azure_devops_project',
    'azure_devops_account_has_project',
    'azure_devops_work_item',
    'azure_devops_project_has_work_item',
    'azure_devops_user', 
    'azure_devops_account_has_user',
    'azure_devops_team',
    'azure_devops_account_has_team', 
    'azure_devops_team_has_user', 
    'azure_devops_project_has_team',
  ],
  "numCollectedEntities": 9,
  "numCollectedRelationships": 13,
}
`;
